*: rewrite coreos-metadata in rust

sdemos opened the pull request:
we've been reviewing it as we go and did a review of the project as a whole as well.

sdemos committed:
*: add cargo project scaffolding

this adds the basic boilerplate for a rust/cargo project.

sdemos committed:
cargo: added clap and slog

these are two dependencies which are needed to start development on the
new version of coreos-metadata.

sdemos committed:
main: simple cli mimicking golang coreos-metadata

this implementation just implements the command-line tool. it is mostly
not implemented still.

sdemos committed:
main: explicitly print configuration error

this is for easier debugging of the cli.

sdemos committed:
main: added metadata fetching structure

this just adds a very simple first pass at how providers will be
structured. this is subject to change in future commits.

sdemos committed:
main: call metadata writing functions

add basic functions for the metadata to actually output its information
in a predictable way for all providers. the functions aren't actually
implemented yet, but they are called by the main function.

sdemos committed:
main: added implementations for metadata writers

this adds a relatively simple implementation of the functions that
output the metadata for all providers.

sdemos committed:
*: break out functionality into multiple files

this provides better structure for the functionality so that I'm not
just working in one giant main file the whole time.

sdemos committed:
network: write network interface/device units

this writes out network information gathered from the provider into
network unit files.

sdemos committed:
macros: added more comprehensive error handling

we don't want to deal with all the error handling boilerplate all over
the place, so use automatically formatting macros.

sdemos committed:
*: convert to new error handling macros

this reduces a lot of the boilerplate in our error handling.

sdemos committed:
network: added unit tests

these are basic unit tests, but the cover a lot of stuff. this as still
yet to be tested with an actual provider though.

sdemos committed:
*: restructure library, add metadata builder

the metadata builder allows us to construct metadata types easier
without having to rely on the internals of the struct. also, the library
approach decouples the fetching of metadata and the functionality of the
command line tool.

sdemos committed:
providers: added azure ssh-keys support

this is the first provider to be implemented. this adds all of the
dependencies and the functionality needed to decrypt the ssh keys given
to us by the azure metadata service. the code should be well commented
enough to explain exactly what has to be done to actually get the
information (it's not straightforward).

sdemos committed:
bin: fixed top-level logging output

this fixed some copy-pasta

sdemos committed:
providers: added provider scaffolding

this adds all the providers that we eventually need to implement at the
time of this writing. none of the implementations, just the modules.

sdemos committed:
azure: get metadata service from fabric address

this adds functionality for reading the OPTION_245 flag from the dhcp
lease so that we are definitely talking to the right azure metadata
service instead of a hard-coded one.

sdemos committed:
cargo: update to latest rust-openssl bindings

for a little while, we were using a for of this package with some
additional functionality. that functionality has now been merged
upstream in sfackler/rust-openssl#675 and we can now just point at the
released package.

sdemos committed:
azure: validate fabric version

we need to make sure that the fabric address that we got actually points
to a metadata service, and that that service is compatible with us.

sdemos committed:
retry: added basic http get client logic

doesn't currently have retries, but it does cut down on repeated http
get code.

sdemos committed:
azure: move to new retry client implementation

this commit actually moves the azure provider over to the new retry
module implementation

Commit:
retry: add retry logic

Replaces hyper dependency with reqwest, rewrites src/retry/mod.rs to use
reqwest and implement http backoff and retry logic, and refactors
src/providers/azure/mod.rs to use the new retry logic.

Commit:
src/retry: add raw deserializer

this allows the retry logic to work with services which don't return
structured data, while still providing automatic deserialization for
those who do.

sdemos committed:
*: move to error-chain

this moves from our error macros to the error-chain library, which lets
us implicitely keep a chain of underlying errors for logging purposes.

Commit:
retry: fixed issue where retry recursed forever

if a client used the default max attempts amount, the retry logic would
recurse infinitely.

Commit:
retry: can return empty results

If return_on_404 is set to true, now Some(None) will be returned if a
404 is encountered, Some(T) is returned if a 200 is encountered (and
successfully deserialized), and an Err is returned otherwise.

Commit:
providers/ec2: initial implementation

this implementation gets attributes from the ec2 metadata service. it
also fixes some mutability warnings with the metadata builder.

Commit:
metadata: change add_ssh_key to take vector

this makes working with the builder easier for providers that give
coreos-metadata multiple ssh public keys (ie most of them)

Commit:
providers/ec2: add ssh keys implementation

this gets ssh public keys from the ec2 metadata service

Commit:
lib: allow unused doc comments in errors mod

this is a bug in the current version of error-chain, and will hopefully
be fixed in the future

Commit:
providers/ec2: fix mutability warning

ssh_keys didn't need to be mutable in this function, since ownership is
passed to the add_ssh_keys builder function.

Commit:
providers/gce: initial implementation

this gets all the metadata from the gce metadata service

sdemos committed:
azure: move to an ssh-keys handling library

this library needs to be more powerful than the current implementation,
and is useful to more than just this library in general.

Commit:
util: add strings parsing function

this adds a new util module with a generally useful key lookup function

sdemos committed:
cargo: updated dependencies to newest version

in particular, update to the newest ssh-keys library version, but also
update the other dependencies.

Commit:
providers/digitalocean: initial implementation

this gets all metadata from the digitalocean metadata service.

Commit:
network: add from_string, bonding modes

add MacAddr::from_string and bonding modes, use ipnetwork crate

Commit:
providers/packet: initial implementation

this adds getting all metadata from the packet metadata service

sdemos committed:
cargo: added clippy feature

this lets us easily run clippy with cargo build --features=[lints]

sdemos committed:
*: fixed issues revealed by clippy

this also adds the clippy plugin when the clippy feature is enabled.

Commit:
providers/ec2: refactor fetch_ssh_keys

this simplifies the ec2 fetch_ssh_keys logic.

Commit:
providers/openstack: initial implementation

this implements getting metadata from the openstack metadata service

Commit:
providers/vagrant_virtualbox: implementation

this implements vagrant_virtual box metadata fetching

sdemos committed:
*: use `update-ssh-keys` crate

the `update-ssh-keys` crate takes care of all the expected file locking and
general behavior around the authorized_keys.d folder. use that instead of our
own solution.

also store ssh keys in PublicKey format instead of just strings

sdemos committed:
network: use pnet::util:MacAddr

we used to implement this ourselves. here we delete it and use a library
we already import to do it instead.

sdemos committed:
cargo: move to published openssh-keys package

this makes our dependencies more simple, since we can pull from a
crates.io package instead of the git repository.

sdemos committed:
providers/azure: added attribute support

the original implementation of the azure provider only got ssh-keys. we
need to make sure the library has complete feature parity with the
golang version.

sdemos committed:
cargo: bump dependencies

update to the newest patch versions of all our dependencies, in
particular, openssh-keys

sdemos committed:
providers/azure: only fetch goalstate once

this reduces the number of network calls we need to get the metadata,
and also makes sure we are always working off the same information. also
adds various derivations for structs.

sdemos committed:
providers/azure: wait until dhcp lease is aquired

There was an issue with the golang version where it would sometimes run
before the dhcp lease was actually aquired, so it would fail to get the
ip address (coreos/coreos-metadata#26). Let's not make the same mistake.

I would prefer it if this was a backoff timer of some kind instead of
just looping forever, but for now, we are just going to do what the
golang implementation does.

sdemos committed:
cargo: bump update-ssh-keys version

the update-ssh-keys library was working around a bug in rust-users
relating to user switching. we update to the newest version of that
package now that the upstream bug is resolved.

sdemos committed:
cargo: bump openssh-keys and update-ssh-keys

there were some bug fixes and additional features in the openssh-keys
and update-ssh-keys libraries, so get the latest versions of those
packages.

sdemos committed:
providers/oracle: initial implementation

this uses the oracle metadata service to fetch metadata

lucab committed:
providers/ec2: add HTTP mock tests

this allows us to be more confident about the rewrite. right now, this
is only implemented for some basic assumptions about the behavior of ssh
key fetching for ec2, but the intention is to eventually expand this to
cover all providers over time.

sdemos committed:
util: add key_lookup_reader function

most of the time that a key is looked up, it is from a file. we make a
function that is generic over readers so that we don't have to manually
read the contents of the file everytime we want to get a key from it.

sdemos committed:
util: add dhcp key lookup function

this function removes a lot of the work from providers that need to look
up this information and lets us only fix issues with the logic in one
place.

fixes #19

sdemos committed:
providers/cloudstack: implement network metadata

this gets metadata from the cloudstack network metadata service.

sdemos committed:
retry: implement generic retrying function

This breaks out the retry logic into it's own separate function. This is
a cleaner and simpler organization strategy, and allows us to use the
retry logic on anything that returns a Result, instead of just the
client logic.

sdemos committed:
util: use retry to find dhcp lease

before this commit, finding the dhcp lease would just loop forever until
it found one. now we have backoff logic so we don't hang forever if the
lease doesn't show up, and instead print an error message.

sdemos committed:
cargo: added bin declaration

For now we aren't exposing the library publicly, so don't let cargo try
and pull all that automatically.

sdemos committed:
cargo: add nix and tempdir dependencies

we need nix and tempdir to mount the configdrive used by the cloudstack
provider.

sdemos committed:
providers/cloudstack: implement configdrive

this module gets metadata from the cloudstack configdrive

sdemos committed:
readme: added readme with provider metadata

basically copied from the existing repo. additionally, it has ssh keys
for azure and network configs for packet on the list.

fixes #14

sdemos committed:
*: delete go

we are rewriting this project in rust. the first step is deleting all
the old code.

sdemos committed:
travis: added travis configuration for rust

now we have continuous integration. this builds on stable, beta, and
nightly, allowing for failures on nightly.

arithx commented on CONTRIBUTING.md:
```diff
@@ -37,17 +44,23 @@ This is a rough outline of what a contributor's workflow looks like:
 
+### Coding Style
+
+CoreOS projects written in Go follow a set of style guidelines that we've documented
```
Doesn't seem necessary to link to the Go style guide when the entire projects in Rust.

sdemos replied:
I just copied this from the template project repo without reading it. I'll update it.

sdemos replied:
actually do we have a convenient place to link to for our rust coding style or should I just delete this section? 

/cc @crawford

crawford replied:
We can just recommend that rustfmt be used.

sdemos committed:
contributing: add rustfmt code style guideline

also removes golang code style guidelines

sdemos committed:
tailor: added tailor config

now we can apply our code style guidlines to commit messages
automatically.

arithx commented on CONTRIBUTING.md:
```diff
@@ -44,11 +44,9 @@ This is a rough outline of what a contributor's workflow looks like:
-CoreOS projects written in Go follow a set of style guidelines that we've documented
-[here](https://github.com/coreos/docs/tree/master/golang). Please follow them when
-working on your contributions.
+Code should be formatted according to the output of [rustfmt](https://github.com/rust-lang-nursery/rustfmt)
```
Can we add a `rustfmt` call to the Travis job similar to how we run `gofmt` on other projects?

sdemos replied:
I'm going to add this in the future to both coreos-metadata and update-ssh-keys.

sdemos committed:
cargo: update dependencies

make sure we are using the latest of all our dependent libraries

dm0- commented on src/util/mod.rs:
```diff
@@ -0,0 +1,107 @@
+    Ok(None)
+}
+
+pub fn key_lookup(delim: char, key: &str, contents: &str) -> Option<String> {
```
Building this results in warnings:

```
warning: function is never used: `key_lookup`
  --> src/util/mod.rs:51:1
   |
51 | / pub fn key_lookup(delim: char, key: &str, contents: &str) -> Option<String> {
52 | |     for l in contents.lines() {
53 | |         if let Some(v) = key_lookup_line(delim, key, l) {
54 | |             return Some(v)
...  |
57 | |     None
58 | | }
   | |_^
   |
   = note: #[warn(dead_code)] on by default
```

lucab replied:
There are a few other clippy warnings as well. I think we can tackle that after merging and setting up travis-ci for it as well. I already did for the other crates we have, and I'm planning to do it here as well.

dm0- commented on Cargo.toml:
```diff
@@ -0,0 +1,45 @@
+[package]
+name = "coreos-metadata"
+version = "0.1.0"
```
So it's mentioned here and not just in the overlay PR, this should also be updated to `1.0.0`.

sdemos committed:
cargo: bump package version to 1.0.0

this is a major update over the previous version of coreos-metadata

sdemos committed:
cargo: update dependencies

in particular, update update-ssh-keys to the latest version

sdemos committed:
cargo: update update-ssh-keys dependency

now that update-ssh-keys has been merged into the upstream
coreos/update-ssh-keys repository, we have to point there for the latest
updates. also, we pin ourselves to a tag in that repo so that our
dependencies are more predictable.
